File: languages/scala/polymorphism.scala
// polymorphism: the dynamic-dispatch / virtual-call-overhead axis of the suite. Build N=10000
// objects of K=6 concrete types in an unpredictable LCG order (the call site is MEGAMORPHIC, so
// the JIT cannot devirtualize / inline-cache it), then fold an accumulator through all of them M
// times: acc = obj.apply(acc). Every type shares the same fields a,b,c but has its OWN apply()
// formula; which one runs is resolved at RUNTIME by the object's concrete type. The acc threads
// through every call (a strict data dependency), so the N*M dispatches cannot be hoisted.
//
// Scala uses idiomatic runtime polymorphism: a trait with an abstract method + six final
// subclasses, held in an Array[Apply] of the base type -> a real virtual (interface) dispatch per
// object. NOT a type tag + match (that would measure a branch, which is forbidden by the spec).
// All integer, mod 1e9+7; checksum = the final accumulator.
object Polymorphism {
  final val P = 1000000007L
  final val N = 10000
  final val K = 6

  // The "vtable": one abstract method, six distinct per-type bodies. Distinct large multipliers
  // keep the per-pass composition chaotic (never a fixed point), so the checksum depends on M.
  abstract class Apply(val a: Long, val b: Long, val c: Long) {
    def apply(x: Long): Long
  }
  final class T0(a: Long, b: Long, c: Long) extends Apply(a, b, c) {
    def apply(x: Long): Long = (x * 1000003L + a) % P
  }
  final class T1(a: Long, b: Long, c: Long) extends Apply(a, b, c) {
    def apply(x: Long): Long = (x * 998273L + b) % P
  }
  final class T2(a: Long, b: Long, c: Long) extends Apply(a, b, c) {
    def apply(x: Long): Long = (x * 999983L + c) % P
  }
  final class T3(a: Long, b: Long, c: Long) extends Apply(a, b, c) {
    def apply(x: Long): Long = (x * 997879L + a + b) % P
  }
  final class T4(a: Long, b: Long, c: Long) extends Apply(a, b, c) {
    def apply(x: Long): Long = (x * 996323L + b * c) % P
  }
  final class T5(a: Long, b: Long, c: Long) extends Apply(a, b, c) {
    def apply(x: Long): Long = (x * 995369L + a + c) % P
  }

  // glibc-style LCG (constants exceed Int range -> Long arithmetic); type from the HIGH bits
  // (the low bits correlate), fields mod 1000.
  def lcg(s: Long): Long = (s * 1103515245L + 12345L) & 0x7fffffffL

  def run(m: Int): Long = {
    val objs = new Array[Apply](N)
    var s = 42L
    var i = 0
    while (i < N) {
      s = lcg(s); val t = ((s >> 16) % K).toInt   // type from HIGH bits; all K used -> megamorphic
      s = lcg(s); val a = s % 1000
      s = lcg(s); val b = s % 1000
      s = lcg(s); val c = s % 1000
      objs(i) = (t: @annotation.switch) match {
        case 0 => new T0(a, b, c)
        case 1 => new T1(a, b, c)
        case 2 => new T2(a, b, c)
        case 3 => new T3(a, b, c)
        case 4 => new T4(a, b, c)
        case _ => new T5(a, b, c)
      }
      i += 1
    }
    var acc = 1L
    var pass = 0
    while (pass < m) {
      var j = 0
      while (j < N) { acc = objs(j).apply(acc); j += 1 }   // DYNAMIC dispatch (virtual call)
      pass += 1
    }
    acc
  }

  def main(args: Array[String]): Unit = {
    val m = if (args.nonEmpty) args(0).toInt else 50
    println(run(m))
    println(s"polymorphism($m)")
  }
}

File: languages/rust/polymorphism.rs
// polymorphism: the dynamic-dispatch / virtual-call-overhead axis. Build N objects of K=6 distinct
// concrete types, mixed in an unpredictable LCG order so the call site is MEGAMORPHIC (K=6 > a
// typical polymorphic inline cache -> no devirtualization). Then fold an accumulator through all of
// them M times: acc = obj.apply(acc). Each type has the SAME fields a,b,c but its OWN apply()
// formula, resolved at RUNTIME from the object's type. acc threads through every call (a strict data
// dependency) and the multipliers are large+distinct, so the per-pass map never reaches a fixed
// point: exactly N*M real dispatches happen and the checksum depends on M.
//
// Rust's idiomatic runtime polymorphism is the TRAIT OBJECT: a `dyn Apply` behind a `Box` carries a
// vtable, and `objs[i].apply(acc)` is a dynamic (indirect) call through that vtable. NOT an
// enum+match (that would be a source-level type tag + branch, which the README forbids).
//
// Every type stores all three fields a,b,c but each apply() reads only a subset (e.g. T0 ignores
// b,c) - faithful to the C/Python references, where each transform uses a different subset. The
// unused fields are intentional, so silence the dead_code warning rather than dropping the field.
#![allow(dead_code)]

use std::env;

const P: i64 = 1000000007;
const N: usize = 10000;
const K: i64 = 6;

// The "virtual method" interface. Six concrete structs implement it; the object stores its type only
// implicitly, in the vtable of the `dyn Apply` it is boxed behind.
trait Apply {
    fn apply(&self, x: i64) -> i64;
}

// All six types share the same three integer fields; only the apply() body differs. The fields are
// duplicated per struct (rather than composed) so each is a genuinely distinct concrete type with
// its own vtable -> the call site sees six targets and stays megamorphic.
struct T0 {
    a: i64,
    b: i64,
    c: i64,
}
struct T1 {
    a: i64,
    b: i64,
    c: i64,
}
struct T2 {
    a: i64,
    b: i64,
    c: i64,
}
struct T3 {
    a: i64,
    b: i64,
    c: i64,
}
struct T4 {
    a: i64,
    b: i64,
    c: i64,
}
struct T5 {
    a: i64,
    b: i64,
    c: i64,
}

// Six distinct per-type transforms (the "virtual method" bodies), each mod P. Distinct large
// multipliers keep acc chaotic so the composition never reaches a fixed point. The products fit in
// i64 (x < P < 2^30, mult < 2^20 -> x*mult < 2^50; b*c < 10^6), but use wrapping_* to make the
// intent explicit and match the suite's LCG style.
impl Apply for T0 {
    fn apply(&self, x: i64) -> i64 {
        (x.wrapping_mul(1000003).wrapping_add(self.a)) % P
    }
}
impl Apply for T1 {
    fn apply(&self, x: i64) -> i64 {
        (x.wrapping_mul(998273).wrapping_add(self.b)) % P
    }
}
impl Apply for T2 {
    fn apply(&self, x: i64) -> i64 {
        (x.wrapping_mul(999983).wrapping_add(self.c)) % P
    }
}
impl Apply for T3 {
    fn apply(&self, x: i64) -> i64 {
        (x.wrapping_mul(997879).wrapping_add(self.a).wrapping_add(self.b)) % P
    }
}
impl Apply for T4 {
    fn apply(&self, x: i64) -> i64 {
        (x.wrapping_mul(996323).wrapping_add(self.b.wrapping_mul(self.c))) % P
    }
}
impl Apply for T5 {
    fn apply(&self, x: i64) -> i64 {
        (x.wrapping_mul(995369).wrapping_add(self.a).wrapping_add(self.c)) % P
    }
}

fn lcg(s: i64) -> i64 {
    (s.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7fffffff
}

fn main() {
    let m: i64 = env::args().nth(1).and_then(|s| s.parse().ok()).unwrap_or(50);

    let mut objs: Vec<Box<dyn Apply>> = Vec::with_capacity(N);
    let mut s: i64 = 42;
    for _ in 0..N {
        s = lcg(s);
        let t = (s >> 16) % K; // type from HIGH bits (LCG low bits correlate); all K used
        s = lcg(s);
        let a = s % 1000;
        s = lcg(s);
        let b = s % 1000;
        s = lcg(s);
        let c = s % 1000;
        // Construct the concrete type chosen at runtime, erased to `dyn Apply` (its vtable is the
        // object's runtime type). Objects stay in LCG generation order -> megamorphic.
        let obj: Box<dyn Apply> = match t {
            0 => Box::new(T0 { a, b, c }),
            1 => Box::new(T1 { a, b, c }),
            2 => Box::new(T2 { a, b, c }),
            3 => Box::new(T3 { a, b, c }),
            4 => Box::new(T4 { a, b, c }),
            _ => Box::new(T5 { a, b, c }),
        };
        objs.push(obj);
    }

    let mut acc: i64 = 1;
    for _ in 0..m {
        for obj in &objs {
            acc = obj.apply(acc); // DYNAMIC dispatch (indirect call through the trait-object vtable)
        }
    }

    println!("{}", acc);
    println!("polymorphism({})", m);
}
